Udf support targeting issue #52

imarios opened the pull request:
Udf support targeting issue #52. More details in #52.

imarios commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,80 @@
+package frameless.functions
```
@OlivierBlanvillain This is the main file where the UDF are defined.

imarios replied:
good catch! by the way, I was always wondering what the difference of the two approaches is ... ?

imarios commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,227 @@
+package frameless.functions
```
@OlivierBlanvillain here are all the tests for UDF.

kanterov replied:
The same comment about
```scala
package frameless
package functions
```

imarios commented on dataset/src/main/scala/frameless/TypedDataset.scala:
```diff
@@ -225,24 +242,167 @@ class TypedDataset[T](val dataset: Dataset[T])(implicit val encoder: TypedEncode
 
+  import scala.reflect.runtime.universe.TypeTag
+
+  def makeUDF[A: TypeTag, R: TypeTag](f: A => R)(implicit aen: TypedEncoder[A], ren: TypedEncoder[R]): TypedColumn[T, A] => TypedColumn[T, R] = functions.udfs.udf[T, A, R](f)
```
@OlivierBlanvillain the makeUDF essentially lifts a UDF to the context of the enclosing TypedDatased.

imarios commented:
Rebased! @OlivierBlanvillain

kanterov commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,80 @@
+     reng: TypedEncoder[R]):
+  (TypedColumn[T, A] => TypedColumn[T, R]) =
+    (u: TypedColumn[T, A]) => {
+      val uf = sudf(f)
```
This is strange that we need `TypeTag`. As I can see from `udf` in Spark it uses `TypeTag` for
```scala
    val inputTypes = Try(ScalaReflection.schemaFor(typeTag[A1]).dataType :: Nil).toOption
```

It wouldn't use `TypedEncoder`, to make use of frameless encoders we have to use `UserDefinedFunction` directly, that takes a list of input types.

imarios replied:
not sure what we mean here, can you elaborate more on this?

kanterov replied:
Vanilla Spark uses reflection to build encoders, that's why it needs `TypeTag`, frameless uses `TypedEncoder`, so we shouldn't need `TypeTag`. We can't reuse `org.apache.spark.sql.functions.udf` because it relies on reflection, but there is `UserDefinedFunction` expression that takes a list of input types that correspond to `TypedEncoder[A..Z].targetDataType`

imarios replied:
oh, cool! Ok, let me get rid of the TypeTags and work directly with the UserDefinedFunction. Thanks!

kanterov replied:
It shows one of the problems with our tests, probably we should also test against something that is only encodable by frameless :)

kanterov commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,80 @@
+      new TypedColumn[T,R](x)
+    }
+
+  def udf[T, A1: TypeTag, A2: TypeTag, A3: TypeTag, A4: TypeTag, R: TypeTag](
```
What about scrapping boilerplate with shapeless like it was done for `selectMany`?

kanterov commented on dataset/src/main/scala/frameless/functions/package.scala:
```diff
@@ -2,4 +2,5 @@ package frameless
 
 package object functions {
   object aggregate extends AggregateFunctions
+  object udfs extends Udf
```
What about extending `functions` with `Udf`? It looks unusual to import `frameless.function.udfs.udf`

kanterov commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,251 @@
+class UdfTest extends TypedDatasetSuite {
+
+  test("one argument udf") {
+    def prop[A, B, C](f1: A => A)(data: Vector[X3[A, B, C]])(
```
What about moving implicits to type parameters? By the way, if we have implicit encoders for `A`,`B`,`C`, it should be derived automatically for `X3`, so no need to require it.

imarios commented on dataset/src/main/scala/frameless/TypedDataset.scala:
```diff
@@ -245,6 +245,32 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
+
+  /** Takes a function from A => R and converts it to a UDF for TypedColumn[T, A] => TypedColumn[T, R].
+    */
+  def makeUDF[A: TypeTag, R: TypeTag](f: A => R)(implicit aen: TypedEncoder[A], ren: TypedEncoder[R]): TypedColumn[T, A] => TypedColumn[T, R] = functions.udfs.udf[T, A, R](f)
```
hmm ... let me try this out

kanterov commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,251 @@
+      val d = data.map(x => f1(x.a))
+
+      dataset21 ?= d
+      dataset22 ?= d
```
The previous line would be dismissed, for conjunction it should be
```scala
(dataset21 ?= d) && (dataset22 ?= d)
```

I'm not sure if we should test our implementation against vanilla Spark, can you elaborate?

imarios replied:
Here we are testing the two ways you can create a udf using our udf function and the makeUDF method from TypedDataset.

imarios commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,251 @@
+      dataset22 ?= d
+    }
+
+    check(forAll(prop[Int, Int, Int, Int, Int]((_: Int) + (_: Int) + (_: Int) + (_: Int) + (_: Int)) _))
```
I did, and it does. I am not sure why I didn't use it at the end. Let me go over this again and it might come back to me.

kanterov commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,80 @@
+     a5enc: TypedEncoder[A5],
+     reng: TypedEncoder[R]):
+  ( (TypedColumn[T, A1],TypedColumn[T, A2],TypedColumn[T, A3],TypedColumn[T, A4],TypedColumn[T, A5]) => TypedColumn[T, R]) =
+    (u1: TypedColumn[T, A1], u2: TypedColumn[T, A2], u3: TypedColumn[T, A3], u4: TypedColumn[T, A4], u5: TypedColumn[T, A5]) => {
```
AFAIK there is no need to specify types here, they would be inferred by the compiler because function signature already has types.

imarios replied:
Let me stip thing out until I it starts complain. I agree, it looks pretty ugly like this and redundant.

imarios commented:
@kanterov great comments! I did a force push with a amended commit, so the review code will not look that great (I just wanted to have one commit for this feature). 

Here is a quick overview of the changes:
- We no longer use Spark udf() function that takes TypeTags and uses reflection. We bypass that by creating a UserDefinedFunction (created inside FramelessInternals). So all TypeTag were removed.
- The makeUDF code inside TypedDataset is has been simplified. We removed the type tags and we use a context bounds instead of implicits for TypedEncoder.
- The test code relies on automatically generated functions by ScalaCheck. We also fixed the `(dataset21 ?= d) && (dataset22 ?= d)` issue.
- All the packages now use the standard frameless package syntax.

imarios commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,194 @@
+class UdfTest extends TypedDatasetSuite {
+
+  test("one argument udf") {
+    def prop[A, B, C](data: Vector[X3[A, B, C]], f1: A => A)(
```
Good catch, I can make it `def prop[A: TypedEncoder, B: TypedEncoder, C: TypedEncoder]`. You think that doing so it will automatically extract an encoding for X3[A,B,C]? Do I need to include something here? I think I tried it and it was complaining.

kanterov replied:
Yes, it should work like this.

imarios replied:
And it did. Thanks.

imarios commented on dataset/src/main/scala/org/apache/spark/sql/FramelessInternals.scala:
```diff
@@ -30,4 +31,6 @@ object FramelessInternals {
     new Dataset(sqlContext, plan, encoder)
   }
+
+  def makeUDF(f: AnyRef, dataType: DataType, inputTypes: Option[Seq[DataType]]) = UserDefinedFunction(f,dataType,inputTypes)
```
Ok, let me try with going directly to the ScalaUDF code.

imarios replied:
All set. Much cleaner for sure without using protected apis.

kanterov commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,194 @@
+    ): Prop = {
+      val dataset = TypedDataset.create(data)
+      val u1 = udf[X3[A, B, C], A, A](f1)
+      val u2 = dataset.makeUDF(f1)
```
It there any difference between `udf` and `makeUDF`? I feel it would be enough to test only `makeUDF`, it would make test code more readable.

imarios replied:
It will actually drop test coverage. Let's just keep it as is.

imarios replied:
This one might be ok to stay as is. It also documents a good contrast between the verbosity of using the udf compared to the makeUDF alternative.

kanterov commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,75 @@
+
+  def udf[T, A1, A2, A3, A4, A5, R](
+    f: (A1,A2,A3,A4,A5) => R)
+    (implicit
```
I was thinking about this, we don't actually need encoders for `A1` because having `u1` already implies we have encoder available as `u1.encoder`, we can do like this:
https://github.com/adelbertc/frameless/blob/master/dataset/src/main/scala/frameless/ops/GroupByOps.scala#L107

imarios replied:
good idea. Did the change.

kanterov commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,75 @@
+  (implicit
+     aenc: TypedEncoder[A],
+     reng: TypedEncoder[R]) =
+    (u: TypedColumn[T, A]) => {
```
I see there was a confusion with my previous comment, I was proposing to make it 
```scala
  def udf[T, A, R](f: A => R)(
     implicit
     aenc: TypedEncoder[A],
     reng: TypedEncoder[R]
  ): (TypedColumn[T, A] => TypedColumn[T, R]) = { 
  u => ...
}
```

General principle is that public methods should always have type signatures, but we don't need to repeat it in body again.

imarios replied:
Got it. Make sense. Changed them all.

kanterov commented:
Please don't squash commits before merge, I find it very difficult to review afterwards.

kanterov commented on dataset/src/test/scala/frameless/functions/UdfTest.scala:
```diff
@@ -0,0 +1,163 @@
+      (dataset21 ?= d) && (dataset22 ?= d)
+    }
+
+    check(forAll(prop[Int, Int, Int] _))
```
I was thinking about this, there is a special case where `A` is case class (for instance `X2[Int, Int]`), and option, for instance `Option[Int]`

kanterov commented on dataset/src/main/scala/frameless/functions/Udf.scala:
```diff
@@ -0,0 +1,97 @@
+    u =>
+      implicit val aenc = u.uencoder
+
+      val scalaUdf = ScalaUDF(f, reng.targetDataType, Seq(u.expr),
```
Sorry for line hunting, but we can save 3 lines per method and improve readability by replacing with:
```scala
  def udf[T, A, R: TypedEncocer](f: A => R): TypedColumn[T, A] => TypedColumn[T, R] = {
    u =>
      val aenc = u.uencoder

      val scalaUdf = ScalaUDF(f, TypedEncoder[R].targetDataType, Seq(u.expr),
        Seq(aenc.targetDataType))
       new TypedColumn[T, R](scalaUdf)
   }
```

AFAIK `implicit val` can be replaced with just a `val`.

kanterov commented:
@OlivierBlanvillain do you have anything to add?

OlivierBlanvillain commented:
no

kanterov commented:
LGTM if there would be a test where `A = X2[Int, Int]`, see https://github.com/adelbertc/frameless/pull/56/files#r85060039

kanterov commented:
LGTM 👍 , let's squash and merge

imarios commented:
All set guys. Thanks for all the comments!

OlivierBlanvillain commented:
@imarios Many thanks for your work! Merging :)
